pkg/api/releases.go

```diff
@@ -43,31 +47,157 @@ func PrintPullRequestsReport(w http.ResponseWriter, req *http.Request, dbClient
+
+	lastFailedPayloadIDs := make([]uint, len(onlyFailedPayloads))
+	for i, lfp := range onlyFailedPayloads {
+		lastFailedPayloadIDs[i] = lfp.ID
+	}
+	logger.WithField("failedPayloadIDs", lastFailedPayloadIDs).Debug("last failed payload IDs")
+
+	// Get the job runs for each of the last failed payloads:
+	jobRuns, err := query.ListPayloadBlockingFailedJobRuns(dbc.DB, lastFailedPayloadIDs)
+	if err != nil {
+		return nil, err
+	}
+	logger.WithField("jobRuns", len(jobRuns)).Debug("loaded latest job runs")
+
+	// Build a map of job run to it's payload, we'll use later for identifying unique payloads:
+	jobRunToPayload := map[uint]string{}
+	for _, jr := range jobRuns {
+		jobRunToPayload[jr.Name] = jr.ReleaseTag.ReleaseTag
+	}
+
+	jobRunIDs := make([]uint, len(jobRuns))
+	for i, jr := range jobRuns {
+		jobRunIDs[i] = jr.Name
+	}
+	logger.WithField("jobRunIDs", jobRunIDs).Debug("got job run IDs")
+
+	// Query all test failures for the given job run IDs:
+	// NOTE: slow query here over our biggest table.
+	failedTests := []models.ProwJobRunTest{}
+	dbc.DB.Preload("Test").Where("prow_job_run_id IN ? AND status = 12", jobRunIDs).Find(&failedTests)
+	logger.WithField("failedTestCount", len(failedTests)).Debug("found failed tests")
+
+	// Iterate all failed tests, aggregate each test name to the number of times we saw it, and the number of unique
+	// jobs we saw it in.
+	testNameToAnalysis := map[string]*apitype.TestFailureAnalysis{}
+	for _, ft := range failedTests {
+		if ft.Test.Name == testgridanalysisapi.OpenShiftTestsName {
+			// Skip the "all tests passed" test we inject, it's not relevant here.
+			continue
+		}
+		if _, ok := testNameToAnalysis[ft.Test.Name]; !ok {
+			testNameToAnalysis[ft.Test.Name] = &apitype.TestFailureAnalysis{
+				Name:           ft.Test.Name,
+				ID:             ft.TestID,
+				FailedJobRuns:  []string{},
+				FailedPayloads: []string{},
+			}
+		}
+		ta := testNameToAnalysis[ft.Test.Name]
+		ta.FailureCount++
+		if !util.StrSliceContains(ta.FailedJobRuns, strconv.Itoa(int(ft.ProwJobRunID))) {
+			ta.FailedJobRuns = append(ta.FailedJobRuns, strconv.Itoa(int(ft.ProwJobRunID)))
+		}
+		jrID := jobRunToPayload[ft.ProwJobRunID]
+		if !util.StrSliceContains(ta.FailedPayloads, jrID) {
+			ta.FailedPayloads = append(ta.FailedPayloads, jrID)
+		}
+	}
+	testFailures := make([]*apitype.TestFailureAnalysis, 0, len(testNameToAnalysis))
+	for _, v := range testNameToAnalysis {
+		testFailures = append(testFailures, v)
+
+		consecPayloadFailsCtr := 0
+		for _, payload := range result.ConsecutiveFailedPayloads {
+			if util.StrSliceContains(v.FailedPayloads, payload) {
+				consecPayloadFailsCtr++
+			}
+		}
+		v.ConsecutiveFailedPayloadsCount = consecPayloadFailsCtr
+		// If at least the last three payloads are in rejected state, and this test is broken in all of them,
+		// flag that we may have a blocker.
+		// TODO: refine these criteria? 75% of last failures?
```

Comment:
Need to refine this, if the last 10 payloads failed and we're in the 3 most recent, it's still likely a payload blocker. Frankly latest 2 is probably a likely blocker. It would also be good to filter out infra/install failures and not let those prevent us calling it a blocker.